codec_union.go

```diff
@@ -226,6 +227,28 @@ func decoderOfResolvedUnion(cfg *frozenConfig, schema Schema) ValDecoder {
 		name := unionResolutionName(schema)
 		if typ, err := cfg.resolver.Type(name); err == nil {
 			decoder := decoderOfType(cfg, schema, typ)
+
+			if typ.Kind() == reflect.Int64 {
+				st := schema.Type()
+				lt := getLogicalType(schema)
+				switch {
+				case st == Int && lt == TimeMillis:
+					typ = reflect2.TypeOf(time.Duration(0))
+
+				case st == Long && lt == TimeMicros:
+					typ = reflect2.TypeOf(time.Duration(0))
+
+				case st == Int && lt == Date:
+					typ = reflect2.TypeOf(time.Time{})
+
+				case st == Long && lt == TimestampMillis:
+					typ = reflect2.TypeOf(time.Time{})
+
+				case st == Long && lt == TimestampMicros:
+					typ = reflect2.TypeOf(time.Time{})
+				}
+			}
+
```

Comment:
In my opinion, this would be better handled in `unionResolutionName`. If you allow for logical type names in this, it would result in a better overall handling of the types.

Reply:
@nrwiersma thanks for your feedback! I'm adding support for logical type names inside `unionResolutionName`, that mean I also need to add those to the resolver right?

```go
	// Register logical types
	r.Register(string(Date), time.Time{})
	r.Register(string(TimestampMillis), time.Time{})
	r.Register(string(TimestampMicros), time.Time{})
	r.Register(string(TimeMillis), time.Duration(0))
	r.Register(string(TimeMicros), time.Duration(0))
	r.Register(string(Duration), time.Duration(0))
```

Reply:
indeed, but i would make the name based on type, so like `long.timestamp-millis` as this will always be unique.